File: app/src/main/java/it/liceoarzignano/bold/safe/SafeLoginDialog.java
package it.liceoarzignano.bold.safe;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.Resources;
import android.support.annotation.NonNull;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;

import it.liceoarzignano.bold.R;

class SafeLoginDialog {
    private final Resources mRes;

    private MaterialDialog mDialog;

    private View mDialogView;
    private EditText mPasswordEditText;
    private TextView mHintTextView;

    /**
     * Log in dialog with password input and hints
     *
     * @param context    used to fetch resources
     * @param isFirstTime adapt ui for first login
     */
    @SuppressLint("InflateParams")
    SafeLoginDialog(Context context, final boolean isFirstTime) {
        mRes = context.getResources();

        LayoutInflater inflater = (LayoutInflater)
                context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        mDialogView = inflater.inflate(R.layout.dialog_safe_unlock, null);

        mPasswordEditText = (EditText) mDialogView.findViewById(R.id.safe_dialog_input);
        mHintTextView = (TextView) mDialogView.findViewById(R.id.safe_dialog_hint_text);

        mHintTextView.setText(mRes.getString(isFirstTime ?
                R.string.safe_dialog_first_hint : R.string.safe_dialog_hint));

        if (isFirstTime) {
            mPasswordEditText.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                }

                @Override
                public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                }

                @Override
                public void afterTextChanged(Editable editable) {
                    String input = mPasswordEditText.getText().toString();
                    int message;
                    if (!input.isEmpty()) {
                        if (input.length() < 8) {
                            message = R.string.safe_dialog_first_hint_low;
                        } else if (input.matches(".*\\d.*") &&
                                input.matches(".*[a-zA-Z].*")) {
                            // Has at least 1 number, 1 letter and more than 8 chars
                            message = R.string.safe_dialog_first_hint_good;
                        } else {
                            // More than 8 chars but only numbers / letters
                            message = R.string.safe_dialog_first_hint_medium;
                        }

                        mHintTextView.setText(mRes.getString(message));
                    }

                }
            });
        }
    }

    /**
     * @return dialog view
     */
    View getView() {
        return mDialogView;
    }

    /**
     * Password input
     */
    @NonNull
    String getInput() {
        return mPasswordEditText.getText().toString();
    }

    /**
     * Build dialog and show it
     *
     * @param mBuilder dialog builder
     */
    void build(MaterialDialog.Builder mBuilder) {
        mDialog = mBuilder.build();
        mDialog.show();
    }

    /**
     * Dismiss dialog but keep this alive
     */
    void dismiss() {
        mDialog.dismiss();
    }

    /**
     * Destroy all the unneeded things,
     * this way we'll make sure input
     * is not stored in memory anymore
     */
    void destroy() {
        mHintTextView = null;
        mPasswordEditText = null;
        mDialogView = null;
        mDialog = null;
    }
}

File: app/src/main/kotlin/it/liceoarzignano/bold/safe/SafeLoginDialog.kt
package it.liceoarzignano.bold.safe

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.support.annotation.StringRes
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import com.afollestad.materialdialogs.MaterialDialog
import it.liceoarzignano.bold.R

@SuppressLint("InflateParams")
internal class SafeLoginDialog constructor(context: Context, private val mFirstTime: Boolean,
                                           hasFingerprint: Boolean,
                                           startFpListener: (calback: Callback,
                                                             ImageView) -> Unit) {
    private var mDialog: MaterialDialog? = null
    private var mView: View
    private var mPasswordEditText: EditText
    private var mFingerprintindicator: ImageView
    private var mHintTextView: TextView
    private val mRes = context.resources

    init {
        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        mView = inflater.inflate(R.layout.dialog_safe_unlock, null)
        mPasswordEditText = mView.findViewById(R.id.safe_dialog_input)
        mFingerprintindicator = mView.findViewById(R.id.safe_dialog_fingerprint_auth)
        mHintTextView = mView.findViewById(R.id.safe_dialog_hint_text)
        mHintTextView.text = mRes.getString(when {
                    mFirstTime -> R.string.safe_dialog_first_hint
                    hasFingerprint -> R.string.safe_dialog_hint_fingerprint
                    else -> R.string.safe_dialog_hint
                })

        if (mFirstTime) {
            mPasswordEditText.addTextChangedListener(listener)
        } else if (hasFingerprint) {
            mFingerprintindicator.visibility = View.VISIBLE
            startFpListener(object : Callback {
                override fun dismiss() {
                    this@SafeLoginDialog.dismiss()
                }
            }, mFingerprintindicator)
        }
    }

    val input: String get() = mPasswordEditText.text.toString()

    private val listener: TextWatcher
        get() = object : TextWatcher {
            override fun beforeTextChanged(c: CharSequence, i: Int, i1: Int, i2: Int) = Unit
            override fun onTextChanged(c: CharSequence, i: Int, i1: Int, i2: Int) = Unit

            override fun afterTextChanged(e: Editable) {
                val input = mPasswordEditText.text.toString()
                @StringRes val message: Int

                if (input.isEmpty()) {
                    return
                }

                message = when {
                    input.length < 8 -> R.string.safe_dialog_first_hint_low
                    // Has at least 1 number, 1 letter and more than 8 chars
                    isPasswordGood(input) -> R.string.safe_dialog_first_hint_good
                    // More than 8 chars but only numbers / letters
                    else -> R.string.safe_dialog_first_hint_medium
                }
                mHintTextView.text = mRes.getString(message)
            }
        }

    private fun isPasswordGood(input: String) =
            input.matches(".*\\d.*".toRegex()) && input.matches(".*[a-zA-Z].*".toRegex())

    fun build(activity: Activity, onPositive: (SafeLoginDialog) -> Unit, onNegative: () -> Unit) {
        val builder = MaterialDialog.Builder(activity)
                .customView(mView, false)
                .canceledOnTouchOutside(false)
                .positiveText(
                        if (mFirstTime) R.string.safe_dialog_first_positive
                        else R.string.safe_dialog_positive)
                .onPositive { _, _ -> onPositive(this) }
                .negativeText(android.R.string.cancel)
                .onNegative { _, _ -> onNegative() }

        mDialog = builder.build()
        mDialog?.show()
    }

    fun dismiss() = mDialog?.dismiss()

    internal interface Callback {
        fun dismiss()
    }
}
